File: sdks/sdk-go/sigilum/identity_store.go
package sigilum

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"
)

const (
	identityRecordVersion = 1
	certificateVersion    = 1
	identitiesDirName     = "identities"
)

var namespacePattern = regexp.MustCompile(`^[a-z0-9][a-z0-9-]{1,62}[a-z0-9]$`)

func DefaultSigilumHome() string {
	home, err := os.UserHomeDir()
	if err != nil {
		return ".sigilum"
	}
	return filepath.Join(home, ".sigilum")
}

func normalizeNamespace(raw string) (string, error) {
	namespace := strings.ToLower(strings.TrimSpace(raw))
	if namespace == "" {
		return "", errors.New("namespace is required")
	}
	if !namespacePattern.MatchString(namespace) {
		return "", errors.New("namespace must match ^[a-z0-9][a-z0-9-]{1,62}[a-z0-9]$ (3-64 chars, lowercase)")
	}
	return namespace, nil
}

func getHomeDir(explicit string) string {
	if strings.TrimSpace(explicit) != "" {
		return explicit
	}
	if env := os.Getenv("SIGILUM_HOME"); strings.TrimSpace(env) != "" {
		return env
	}
	return DefaultSigilumHome()
}

func identityDir(homeDir, namespace string) string {
	return filepath.Join(homeDir, identitiesDirName, namespace)
}

func identityPath(homeDir, namespace string) string {
	return filepath.Join(identityDir(homeDir, namespace), "identity.json")
}

func makeDID(namespace string) string {
	return "did:sigilum:" + namespace
}

func fingerprint(publicKey []byte) string {
	sum := sha256.Sum256(publicKey)
	return hex.EncodeToString(sum[:8])
}

func makeKeyID(did string, publicKey []byte) string {
	return did + "#ed25519-" + fingerprint(publicKey)
}

func nowISO() string {
	return time.Now().UTC().Format(time.RFC3339)
}

func certificatePayload(cert SigilumCertificate) []byte {
	expiresAt := ""
	if cert.ExpiresAt != nil {
		expiresAt = *cert.ExpiresAt
	}
	payload := []string{
		"sigilum-certificate-v1",
		"namespace:" + cert.Namespace,
		"did:" + cert.DID,
		"key-id:" + cert.KeyID,
		"public-key:" + cert.PublicKey,
		"issued-at:" + cert.IssuedAt,
		"expires-at:" + expiresAt,
	}
	return []byte(strings.Join(payload, "\n"))
}

func toBase64URL(value []byte) string {
	return base64.RawURLEncoding.EncodeToString(value)
}

func fromBase64URL(value string) ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(value)
}

func publicKeyFromEncoded(encoded string) ([]byte, error) {
	if !strings.HasPrefix(encoded, "ed25519:") {
		return nil, fmt.Errorf("unsupported public key format: %s", encoded)
	}
	decoded, err := base64.StdEncoding.DecodeString(strings.TrimPrefix(encoded, "ed25519:"))
	if err != nil {
		return nil, err
	}
	if len(decoded) != ed25519.PublicKeySize {
		return nil, fmt.Errorf("invalid public key length: %d", len(decoded))
	}
	return decoded, nil
}

func VerifyCertificate(cert SigilumCertificate) bool {
	if cert.Version != certificateVersion {
		return false
	}
	if cert.Proof.Alg != "ed25519" {
		return false
	}
	publicKey, err := publicKeyFromEncoded(cert.PublicKey)
	if err != nil {
		return false
	}
	signature, err := fromBase64URL(cert.Proof.Sig)
	if err != nil {
		return false
	}
	return ed25519.Verify(ed25519.PublicKey(publicKey), certificatePayload(cert), signature)
}

type identityRecord struct {
	Version     int               `json:"version"`
	Namespace   string            `json:"namespace"`
	DID         string            `json:"did"`
	KeyID       string            `json:"keyId"`
	PublicKey   string            `json:"publicKey"`
	PrivateKey  string            `json:"privateKey"`
	Certificate SigilumCertificate `json:"certificate"`
	CreatedAt   string            `json:"createdAt"`
	UpdatedAt   string            `json:"updatedAt"`
}

func createIdentityRecord(namespace string) (*identityRecord, error) {
	publicKey, privateKey, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, err
	}
	seed := privateKey.Seed()
	publicKeyBase64 := base64.StdEncoding.EncodeToString(publicKey)
	did := makeDID(namespace)
	keyID := makeKeyID(did, publicKey)
	now := nowISO()

	cert := SigilumCertificate{
		Version:   certificateVersion,
		Namespace: namespace,
		DID:       did,
		KeyID:     keyID,
		PublicKey: "ed25519:" + publicKeyBase64,
		IssuedAt:  now,
		ExpiresAt: nil,
		Proof: SigilumCertificateProof{
			Alg: "ed25519",
			Sig: "",
		},
	}

	signature := ed25519.Sign(privateKey, certificatePayload(cert))
	cert.Proof.Sig = toBase64URL(signature)

	return &identityRecord{
		Version:     identityRecordVersion,
		Namespace:   namespace,
		DID:         did,
		KeyID:       keyID,
		PublicKey:   "ed25519:" + publicKeyBase64,
		PrivateKey:  base64.StdEncoding.EncodeToString(seed),
		Certificate: cert,
		CreatedAt:   now,
		UpdatedAt:   now,
	}, nil
}

func writeIdentityRecord(homeDir, namespace string, record *identityRecord) (string, error) {
	dir := identityDir(homeDir, namespace)
	path := identityPath(homeDir, namespace)

	if err := os.MkdirAll(dir, 0o700); err != nil {
		return "", err
	}

	bytes, err := json.MarshalIndent(record, "", "  ")
	if err != nil {
		return "", err
	}
	bytes = append(bytes, '\n')

	if err := os.WriteFile(path, bytes, 0o600); err != nil {
		return "", err
	}
	return path, nil
}

func readIdentityRecord(path string) (*identityRecord, error) {
	bytes, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var record identityRecord
	if err := json.Unmarshal(bytes, &record); err != nil {
		return nil, fmt.Errorf("failed to parse identity file %s: %w", path, err)
	}
	if record.Version != identityRecordVersion {
		return nil, fmt.Errorf("unsupported identity version: %d", record.Version)
	}
	if record.Namespace == "" || record.DID == "" || record.KeyID == "" || record.PublicKey == "" || record.PrivateKey == "" {
		return nil, fmt.Errorf("identity file %s is missing required fields", path)
	}
	return &record, nil
}

func ListNamespaces(explicitHomeDir string) ([]string, error) {
	home := getHomeDir(explicitHomeDir)
	root := filepath.Join(home, identitiesDirName)

	entries, err := os.ReadDir(root)
	if err != nil {
		if os.IsNotExist(err) {
			return []string{}, nil
		}
		return nil, err
	}

	namespaces := make([]string, 0, len(entries))
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		identity := filepath.Join(root, entry.Name(), "identity.json")
		if _, err := os.Stat(identity); err == nil {
			namespaces = append(namespaces, entry.Name())
		}
	}
	sort.Strings(namespaces)
	return namespaces, nil
}

func resolveNamespace(explicitNamespace, explicitHomeDir string) (string, error) {
	if strings.TrimSpace(explicitNamespace) != "" {
		return normalizeNamespace(explicitNamespace)
	}
	if env := strings.TrimSpace(os.Getenv("SIGILUM_NAMESPACE")); env != "" {
		return normalizeNamespace(env)
	}

	namespaces, err := ListNamespaces(explicitHomeDir)
	if err != nil {
		return "", err
	}
	if len(namespaces) == 1 {
		return namespaces[0], nil
	}
	if len(namespaces) == 0 {
		return "", errors.New("no Sigilum identity found. Run `sigilum init <namespace>` first")
	}
	return "", fmt.Errorf("multiple identities found (%s). Pass namespace explicitly or set SIGILUM_NAMESPACE", strings.Join(namespaces, ", "))
}

func LoadIdentity(options LoadIdentityOptions) (SigilumIdentity, error) {
	home := getHomeDir(options.HomeDir)
	namespace, err := resolveNamespace(options.Namespace, home)
	if err != nil {
		return SigilumIdentity{}, err
	}

	path := identityPath(home, namespace)
	record, err := readIdentityRecord(path)
	if err != nil {
		if os.IsNotExist(err) {
			return SigilumIdentity{}, fmt.Errorf("Sigilum identity not found for namespace %q at %s. Run `sigilum init %s` first", namespace, path, namespace)
		}
		return SigilumIdentity{}, err
	}

	seed, err := base64.StdEncoding.DecodeString(record.PrivateKey)
	if err != nil {
		return SigilumIdentity{}, fmt.Errorf("invalid private key encoding: %w", err)
	}
	if len(seed) != ed25519.SeedSize {
		return SigilumIdentity{}, fmt.Errorf("invalid private key length: %d", len(seed))
	}

	privateKey := ed25519.NewKeyFromSeed(seed)
	publicKey := privateKey.Public().(ed25519.PublicKey)
	expectedPublic := "ed25519:" + base64.StdEncoding.EncodeToString(publicKey)
	if expectedPublic != record.PublicKey {
		return SigilumIdentity{}, errors.New("public key mismatch in identity file")
	}

	if !VerifyCertificate(record.Certificate) {
		return SigilumIdentity{}, errors.New("identity certificate verification failed")
	}
	if record.Certificate.Namespace != namespace ||
		record.Certificate.DID != record.DID ||
		record.Certificate.KeyID != record.KeyID ||
		record.Certificate.PublicKey != record.PublicKey {
		return SigilumIdentity{}, errors.New("certificate and identity record do not match")
	}

	return SigilumIdentity{
		Namespace:    namespace,
		DID:          record.DID,
		KeyID:        record.KeyID,
		PublicKey:    record.PublicKey,
		PrivateKey:   seed,
		Certificate:  record.Certificate,
		HomeDir:      home,
		IdentityPath: path,
	}, nil
}

func InitIdentity(options InitIdentityOptions) (InitIdentityResult, error) {
	namespace, err := normalizeNamespace(options.Namespace)
	if err != nil {
		return InitIdentityResult{}, err
	}
	home := getHomeDir(options.HomeDir)
	path := identityPath(home, namespace)

	if _, err := os.Stat(path); err == nil && !options.Force {
		identity, err := LoadIdentity(LoadIdentityOptions{Namespace: namespace, HomeDir: home})
		if err != nil {
			return InitIdentityResult{}, err
		}
		return InitIdentityResult{
			Namespace:    identity.Namespace,
			DID:          identity.DID,
			KeyID:        identity.KeyID,
			PublicKey:    identity.PublicKey,
			Created:      false,
			HomeDir:      home,
			IdentityPath: path,
		}, nil
	}

	record, err := createIdentityRecord(namespace)
	if err != nil {
		return InitIdentityResult{}, err
	}
	finalPath, err := writeIdentityRecord(home, namespace, record)
	if err != nil {
		return InitIdentityResult{}, err
	}

	return InitIdentityResult{
		Namespace:    namespace,
		DID:          record.DID,
		KeyID:        record.KeyID,
		PublicKey:    record.PublicKey,
		Created:      true,
		HomeDir:      home,
		IdentityPath: finalPath,
	}, nil
}

func GetNamespaceAPIBase(apiBaseURL, namespace string) string {
	base := strings.TrimRight(apiBaseURL, "/")
	return base + "/v1/namespaces/" + url.PathEscape(namespace)
}

File: sdks/sdk-ts/src/identity-store.ts
import "./setup.js";
import * as ed from "@noble/ed25519";
import { sha256 } from "@noble/hashes/sha2.js";
import { bytesToHex } from "@noble/hashes/utils.js";
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import type {
  InitIdentityOptions,
  InitIdentityResult,
  LoadIdentityOptions,
  SigilumCertificate,
  SigilumIdentity,
  StoredIdentityRecord,
} from "./types.js";

const IDENTITY_RECORD_VERSION = 1;
const CERTIFICATE_VERSION = 1;
const IDENTITIES_DIR = "identities";

export const DEFAULT_SIGILUM_HOME = path.join(os.homedir(), ".sigilum");

function normalizeNamespace(raw: string): string {
  const namespace = raw.trim().toLowerCase();
  if (!namespace) {
    throw new Error("Namespace is required");
  }
  if (!/^[a-z0-9][a-z0-9-]{1,62}[a-z0-9]$/.test(namespace)) {
    throw new Error(
      "Namespace must match ^[a-z0-9][a-z0-9-]{1,62}[a-z0-9]$ (3-64 chars, lowercase)",
    );
  }
  return namespace;
}

function toBase64Url(bytes: Uint8Array): string {
  return Buffer.from(bytes)
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/g, "");
}

function fromBase64Url(value: string): Uint8Array {
  const normalized = value.replace(/-/g, "+").replace(/_/g, "/");
  const pad = normalized.length % 4;
  const padded = normalized + (pad === 0 ? "" : "=".repeat(4 - pad));
  return new Uint8Array(Buffer.from(padded, "base64"));
}

function getHomeDir(explicitHomeDir?: string): string {
  return explicitHomeDir ?? process.env.SIGILUM_HOME ?? DEFAULT_SIGILUM_HOME;
}

function getIdentityDir(homeDir: string, namespace: string): string {
  return path.join(homeDir, IDENTITIES_DIR, namespace);
}

function getIdentityPath(homeDir: string, namespace: string): string {
  return path.join(getIdentityDir(homeDir, namespace), "identity.json");
}

function makeDid(namespace: string): string {
  return `did:sigilum:${namespace}`;
}

function makeFingerprint(publicKey: Uint8Array): string {
  return bytesToHex(sha256(publicKey).slice(0, 8));
}

function makeKeyId(did: string, publicKey: Uint8Array): string {
  return `${did}#ed25519-${makeFingerprint(publicKey)}`;
}

function buildCertificatePayload(cert: Omit<SigilumCertificate, "proof">): string {
  return [
    "sigilum-certificate-v1",
    `namespace:${cert.namespace}`,
    `did:${cert.did}`,
    `key-id:${cert.keyId}`,
    `public-key:${cert.publicKey}`,
    `issued-at:${cert.issuedAt}`,
    `expires-at:${cert.expiresAt ?? ""}`,
  ].join("\n");
}

function createIdentityRecord(namespace: string): StoredIdentityRecord {
  const privateKey = ed.utils.randomSecretKey();
  const publicKey = ed.getPublicKey(privateKey);
  const publicKeyBase64 = Buffer.from(publicKey).toString("base64");
  const did = makeDid(namespace);
  const keyId = makeKeyId(did, publicKey);
  const now = new Date().toISOString();

  const unsignedCertificate: Omit<SigilumCertificate, "proof"> = {
    version: CERTIFICATE_VERSION,
    namespace,
    did,
    keyId,
    publicKey: `ed25519:${publicKeyBase64}`,
    issuedAt: now,
    expiresAt: null,
  };

  const certificatePayload = new TextEncoder().encode(
    buildCertificatePayload(unsignedCertificate),
  );
  const certificateSignature = ed.sign(certificatePayload, privateKey);

  const certificate: SigilumCertificate = {
    ...unsignedCertificate,
    proof: {
      alg: "ed25519",
      sig: toBase64Url(certificateSignature),
    },
  };

  return {
    version: IDENTITY_RECORD_VERSION,
    namespace,
    did,
    keyId,
    publicKey: `ed25519:${publicKeyBase64}`,
    privateKey: Buffer.from(privateKey).toString("base64"),
    certificate,
    createdAt: now,
    updatedAt: now,
  };
}

function writeIdentityRecord(
  homeDir: string,
  namespace: string,
  record: StoredIdentityRecord,
): string {
  const identityDir = getIdentityDir(homeDir, namespace);
  const identityPath = getIdentityPath(homeDir, namespace);
  fs.mkdirSync(identityDir, { recursive: true, mode: 0o700 });
  fs.writeFileSync(identityPath, `${JSON.stringify(record, null, 2)}\n`, {
    mode: 0o600,
  });
  return identityPath;
}

function readIdentityRecord(identityPath: string): StoredIdentityRecord {
  let parsed: unknown;
  try {
    parsed = JSON.parse(fs.readFileSync(identityPath, "utf8"));
  } catch (error) {
    throw new Error(`Failed to read identity file ${identityPath}: ${String(error)}`);
  }

  if (!parsed || typeof parsed !== "object") {
    throw new Error(`Identity file ${identityPath} is invalid`);
  }

  const record = parsed as Partial<StoredIdentityRecord>;
  if (record.version !== IDENTITY_RECORD_VERSION) {
    throw new Error(
      `Unsupported identity version in ${identityPath}: ${String(record.version)}`,
    );
  }

  if (
    !record.namespace
    || !record.did
    || !record.keyId
    || !record.publicKey
    || !record.privateKey
    || !record.certificate
  ) {
    throw new Error(`Identity file ${identityPath} is missing required fields`);
  }

  return record as StoredIdentityRecord;
}

function parsePublicKey(publicKey: string): Uint8Array {
  if (!publicKey.startsWith("ed25519:")) {
    throw new Error(`Unsupported public key format: ${publicKey}`);
  }
  return new Uint8Array(Buffer.from(publicKey.slice("ed25519:".length), "base64"));
}

function resolveNamespace(namespace?: string, homeDir?: string): string {
  if (namespace) {
    return normalizeNamespace(namespace);
  }

  const envNamespace = process.env.SIGILUM_NAMESPACE;
  if (envNamespace) {
    return normalizeNamespace(envNamespace);
  }

  const namespaces = listNamespaces(homeDir);
  if (namespaces.length === 1) {
    const firstNamespace = namespaces[0];
    if (!firstNamespace) {
      throw new Error("Identity lookup failed unexpectedly");
    }
    return firstNamespace;
  }

  if (namespaces.length === 0) {
    throw new Error(
      "No Sigilum identity found. Run `sigilum init <namespace>` first.",
    );
  }

  throw new Error(
    `Multiple identities found (${namespaces.join(", ")}). Pass namespace explicitly or set SIGILUM_NAMESPACE.`,
  );
}

export function verifyCertificate(certificate: SigilumCertificate): boolean {
  if (certificate.version !== CERTIFICATE_VERSION) {
    return false;
  }
  if (certificate.proof.alg !== "ed25519") {
    return false;
  }

  const publicKeyBytes = parsePublicKey(certificate.publicKey);
  const payload = new TextEncoder().encode(
    buildCertificatePayload({
      version: certificate.version,
      namespace: certificate.namespace,
      did: certificate.did,
      keyId: certificate.keyId,
      publicKey: certificate.publicKey,
      issuedAt: certificate.issuedAt,
      expiresAt: certificate.expiresAt,
    }),
  );

  return ed.verify(fromBase64Url(certificate.proof.sig), payload, publicKeyBytes);
}

export function initIdentity(options: InitIdentityOptions): InitIdentityResult {
  const namespace = normalizeNamespace(options.namespace);
  const homeDir = getHomeDir(options.homeDir);
  const identityPath = getIdentityPath(homeDir, namespace);
  const exists = fs.existsSync(identityPath);

  if (exists && !options.force) {
    const identity = loadIdentity({ namespace, homeDir });
    return {
      namespace: identity.namespace,
      did: identity.did,
      keyId: identity.keyId,
      publicKey: identity.publicKey,
      created: false,
      homeDir,
      identityPath,
    };
  }

  const record = createIdentityRecord(namespace);
  const finalPath = writeIdentityRecord(homeDir, namespace, record);

  return {
    namespace,
    did: record.did,
    keyId: record.keyId,
    publicKey: record.publicKey,
    created: true,
    homeDir,
    identityPath: finalPath,
  };
}

export function listNamespaces(explicitHomeDir?: string): string[] {
  const homeDir = getHomeDir(explicitHomeDir);
  const identitiesDir = path.join(homeDir, IDENTITIES_DIR);
  if (!fs.existsSync(identitiesDir)) {
    return [];
  }

  return fs
    .readdirSync(identitiesDir)
    .filter((entry) => {
      const identityPath = path.join(identitiesDir, entry, "identity.json");
      return fs.existsSync(identityPath);
    })
    .sort();
}

export function loadIdentity(options: LoadIdentityOptions = {}): SigilumIdentity {
  const homeDir = getHomeDir(options.homeDir);
  const namespace = resolveNamespace(options.namespace, homeDir);
  const identityPath = getIdentityPath(homeDir, namespace);

  if (!fs.existsSync(identityPath)) {
    throw new Error(
      `Sigilum identity not found for namespace "${namespace}" at ${identityPath}. Run \`sigilum init ${namespace}\` first.`,
    );
  }

  const record = readIdentityRecord(identityPath);

  if (record.namespace !== namespace) {
    throw new Error(
      `Identity namespace mismatch in ${identityPath}: expected ${namespace}, got ${record.namespace}`,
    );
  }

  const privateKey = new Uint8Array(Buffer.from(record.privateKey, "base64"));
  if (privateKey.length !== 32) {
    throw new Error(`Invalid private key length in ${identityPath}`);
  }

  const derivedPublicKey = ed.getPublicKey(privateKey);
  const derivedPublicKeyBase64 = Buffer.from(derivedPublicKey).toString("base64");
  const expectedPublicKey = `ed25519:${derivedPublicKeyBase64}`;
  if (expectedPublicKey !== record.publicKey) {
    throw new Error(`Public key mismatch in ${identityPath}`);
  }

  if (!verifyCertificate(record.certificate)) {
    throw new Error(`Identity certificate verification failed for ${identityPath}`);
  }

  if (
    record.certificate.namespace !== namespace
    || record.certificate.did !== record.did
    || record.certificate.keyId !== record.keyId
    || record.certificate.publicKey !== record.publicKey
  ) {
    throw new Error(`Identity certificate fields do not match record in ${identityPath}`);
  }

  return {
    namespace,
    did: record.did,
    keyId: record.keyId,
    publicKey: record.publicKey,
    privateKey,
    certificate: record.certificate,
    homeDir,
    identityPath,
  };
}

export function getNamespaceApiBase(apiBaseUrl: string, namespace: string): string {
  const base = apiBaseUrl.replace(/\/+$/, "");
  return `${base}/v1/namespaces/${encodeURIComponent(namespace)}`;
}
